march_rqt_gait_generator/src/march_rqt_gait_generator/model/modifiable_joint_trajectory.py

```diff
@@ -23,6 +23,40 @@ def __init__(self, name, limits, setpoints, duration, gait_generator):
 
         self.interpolated_setpoints = self.interpolate_setpoints()
 
+    @classmethod
+    def from_dict(cls, subgait_dict, joint_name, limits, duration):
+        user_defined_setpoints = subgait_dict['setpoints']
+        if user_defined_setpoints:
+            joint_trajectory = subgait_dict['trajectory']
+            setpoints = []
+            for actual_setpoint in user_defined_setpoints:
+                if joint_name in actual_setpoint['joint_names']:
+                    setpoints.append(cls.get_setpoint_at_duration(
+                        joint_trajectory, joint_name, actual_setpoint['time_from_start']))
+
+            return cls(joint_name,
+                       limits,
+                       setpoints,
+                       duration
+                       )
+
+        rospy.logwarn('This subgait has no user defined setpoints.')
+        return cls.super(subgait_dict, joint_name, limits, duration)
+
+    @staticmethod
+    def get_setpoint_at_duration(joint_trajectory, joint_name, duration):
```

Comment:
if this method is only used within this class you should name it `_get_setpoint_at_duration`